perf: adds --no-deploy (when appropriate) to hardhat fork command

mikeshultz opened the pull request:
### What I did

Updated `HardhatForkProvider` to use `--no-deploy` when running `hardhat node --fork [...]` when the `hardhat-deploy` plugin is installed to prevent contract deployments from being automatically run.  

### How I did it

- `HardhatProvider` now loads `package.json` and reads the installed Hardhat plugins (assuming `hardhat-` package prefixes)
- If `enable_hardhat_deployments` is not set (or set to false), and the `hardhat-deploy` plugin is installed, it will add `--no-deploy` option to the `hardhat node [...]`

### How to verify it

```bash
pytest
```

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

mikeshultz commented on README.md:
```diff
@@ -84,6 +84,17 @@ hardhat:
+    ethereum:
+      mainnet:
+        upstream_provider: alchemy
+        no_deploy: false
```
I was thinking about that before but I think the name `no_deploy` makes the most sense since it directly relates to the hardhat option `--no-deploy`.   Though we could ignore that relationship entirely and go with something descriptive like `hardhat_deployments: true`.  Your call.

mikeshultz commented on ape_hardhat/provider.py:
```diff
@@ -198,6 +210,33 @@ def is_connected(self) -> bool:
+        json_path = self.project_folder / "package.json"
+
+        if not json_path.is_file():
+            return PackageJson()
```
Everything is set `Optional` type and seems to work.  Though Pydantic is new to me so I don't know what is convention.

unparalleled-js replied:
I didn't realize! If it works like that by default, no need to specify.

mikeshultz commented on ape_hardhat/provider.py:
```diff
@@ -198,6 +210,33 @@ def is_connected(self) -> bool:
+        plugins: List[str] = []
+
+        def package_is_plugin(package: str) -> bool:
+            return re.search(r"hardhat-[A-Za-z0-9-]+$", package) is not None
```
Have any more info on that?

mikeshultz replied:
Was curious so went looking.  [According to the Python docs](https://docs.python.org/3/howto/regex.html#module-level-functions):

> Under the hood, these functions simply create a pattern object for you and call the appropriate method on it. They also store the compiled object in a cache, so future calls using the same RE won’t need to parse the pattern again and again.

So performance wise it sounds the same to me.  That said, maybe making the compiled pattern a file-level const might be a little cleaner?

unparalleled-js replied:
But look at the last paragraph in the link you sent me:

```
"""
Should you use these module-level functions, or should you get the pattern and call its methods yourself? If you’re accessing a regex within a loop, pre-compiling it will save a few function calls. Outside of loops, there’s not much difference thanks to the internal cache.
"""
```

So I am saying do exactly that since this is getting called in a loop, let's pre-compile it.
But you are right, it is not a optimization because of the caching.

unparalleled-js replied:
> @unparalleled-js so that type check failure doesn't seem related to my PR (I didn't touch pyyaml at all). Is this something you'd want me to address in this PR?

Yeah latest mypy has been causing issues, you can pin in setup.py to fix this issue. That is what we did in Ape core.

mikeshultz replied:
Hm, thanks.  Guess that part didn't register when I read it.  Now I'm wondering if a file-level constant or defining it within the function like your suggestion makes any difference when it comes to the compiler.  I have zero idea, though I'd probably favor the constant even though it's only used in one place.  Let me know if you have a preference.

mikeshultz commented on ape_hardhat/provider.py:
```diff
@@ -99,9 +102,18 @@ def write_if_not_exists(self):
+    version: Optional[str]
+    description: Optional[str]
+    dependencies: Optional[Dict[str, str]]
+    devDependencies: Optional[Dict[str, str]]
```
Since it's a representation of JSON I wonder if camelCase actually makes sense here, but don't want to diverge from your style.  Can do.

unparalleled-js replied:
It is how we handle all of these JSON models in the framework. We use `alias` to point to the camel case and keep the Python in snake case.

mikeshultz commented:
@unparalleled-js so that type check failure doesn't seem related to my PR (I didn't touch pyyaml at all).  Is this something you'd want me to address in this PR?

I was able to get my pre-commit hook working locally  with this change to `.pre-commit-config.yaml` :

    -        additional_dependencies: [types-requests]
    +        additional_dependencies: [types-PyYAML, types-requests]

But I don't think that will fix the GH Action.  Looks like `types-PyYAML` is installed in the job so it's not immediately clear why this would fail.

unparalleled-js commented on ape_hardhat/provider.py:
```diff
@@ -99,9 +102,18 @@ def write_if_not_exists(self):
 
 
+class PackageJson(BaseModel):
+    name: Optional[str]
```
Oh okay, that works!

mikeshultz commented:
@unparalleled-js I think I addressed your feedback.  I also added a couple of changes to `.pre-commit-config.yaml` you should be aware of that fixes some issues I was having using the pre-commit hooks.

mikeshultz commented:
Err, I completely forgot about the config key change.  Will have that up a little later.

mikeshultz commented:
Okay, should be all set now.  Thanks for the feedback.

unparalleled-js commented on ape_hardhat/provider.py:
```diff
@@ -99,9 +103,18 @@ def write_if_not_exists(self):
+    version: Optional[str]
+    description: Optional[str]
+    dependencies: Optional[Dict[str, str]]
+    dev_dependencies: Optional[Dict[str, str]] = Field(None, alias="devDependencies")
```
nbd, but based on what I learned from you, the `None` may do nothing here.

unparalleled-js commented on ape_hardhat/provider.py:
```diff
@@ -198,6 +211,33 @@ def is_connected(self) -> bool:
+        if not json_path.is_file():
+            return PackageJson()
+
+        return PackageJson(**json.loads(json_path.read_text()))
```
```suggestion
        return PackageJson.parse_raw(json_path.read_text())
```

unparalleled-js commented on tests/test_fork_provider.py:
```diff
@@ -156,3 +156,76 @@ def test_get_receipt(connected_mainnet_fork_provider, fork_contract_instance, ow
     assert actual.sender == receipt.sender
+
+
+@pytest.mark.sync
```
Just an FYI - these fork tests currently don't run in CI.

I am wondering if we actually need the `@pytest.mark.sync` for these tests.

If it not actually making provider RPC requests, we maybe can drop this.

mikeshultz replied:
I'm not sure what that is and couldn't immediately find that in pytest docs or conftest.  Will leave this to you.

unparalleled-js reviewed (approved):
Really small feedback that I should have added originally probably.

Thank you so much for your contribution! 

(I may have time to address my own small feedback later today)

mikeshultz commented:
Addressed your other feedback, btw.
